File: Binary Tree/二叉搜索树/450.删除二叉搜索树中的节点.cpp
using namespace std;
/*
 * @lc app=leetcode.cn id=450 lang=cpp
 *
 * [450] 删除二叉搜索树中的节点
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution
{
public:
    TreeNode *deleteNode(TreeNode *root, int key)
    {
        // 没有找到，直接返回空指针
        if (root == nullptr)
        {
            return root;
        }

        if (root->val > key)
        {
            root->left = deleteNode(root->left, key);
        }
        else if (root->val < key)
        {
            root->right = deleteNode(root->right, key);
        }
        else
        {
            // root->val == key 找到这个节点
            // 左右都为空，直接移除这个节点，置为空指针
            if (root->left == nullptr && root->right == nullptr)
            {
                return nullptr;
            }
            // 左子树不为空，右子树为空
            if (root->left != nullptr && root->right == nullptr)
            {
                return root->left;
            }

            // 右子树不为空，左子树为空
            if (root->right != nullptr && root->left == nullptr)
            {
                return root->right;
            }

            // 左右子树都不为空，找右子树最左面的节点
            TreeNode *cur = root->right;
            while (cur->left != nullptr)
            {
                cur = cur->left;
            }
            // 把要删除的节点的左子树作为右子树的最下面节点的左子树进行接入
            // 理解这一句，非常重要
            cur->left = root->left;
            // 右子树作为新的节点进行返回
            root = root->right;

            return root;
        }
        return root;
    }
};
// @lc code=end

File: Binary Tree/二叉搜索树/450.删除二叉搜索树中的节点.ts
/*
 * @lc app=leetcode.cn id=450 lang=typescript
 *
 * [450] 删除二叉搜索树中的节点
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function deleteNode(root: TreeNode | null, key: number): TreeNode | null {
  const findDeleteNode = (node: TreeNode | null): TreeNode | null => {
    if (!node) {
      return null
    }

    if (node.val > key) {
      // 删除节点在左边，上出完后返回新的左子树
      node.left = findDeleteNode(node.left)
    } else if (node.val < key) {
      // 删除节点在右边，上出完后返回新的右子树
      node.right = findDeleteNode(node.right)
    } else {
      // 找到了需要删除的节点
      if (!node.left && !node.right) {
        // 没有左右子节点，直接删除
        node = null
      } else if (node.left && !node.right) {
        // 只有左子节点
        // 左子节点替换为当前节点
        node = node.left
      } else if (!node.left && node.right) {
        // 只有右子节点
        // 右子节点替换为当前节点
        node = node.right
      } else {
        // 左右子节点都不为空
        let leftChild = node.left
        let rightChild = node.right

        // 找到右子树的最左边子节点
        while (rightChild.left) {
          rightChild = rightChild.left
        }

        //把左子树挂到右子树的最左子节点的左边
        rightChild.left = leftChild

        // 右子树替换需要删除的节点
        node = node.right
      }
    }
    // 返回删除完成新的子树
    return node
  }

  root = findDeleteNode(root)

  return root
};
// @lc code=end

